Code/ForceField/ForceField.cpp

```diff
@@ -14,6 +14,80 @@
 #include <RDGeneral/Invariant.h>
 #include <Numerics/Optimizer/BFGSOpt.h>
 
+namespace RDKit {
+namespace ForceFieldsHelper {
+void normalizeAngleDeg(double &angleDeg) {
+  double normFactor = 360.0;
+  if (angleDeg < 0.0)
+    normFactor = -normFactor;
+  angleDeg = fmod(angleDeg, normFactor);
+  if (fabs(angleDeg) > 180.0)
+    angleDeg -= normFactor;
+}
+
+void computeDihedral(const RDGeom::PointPtrVect &pos, unsigned int idx1,
+    unsigned int idx2, unsigned int idx3, unsigned int idx4,
+    double *dihedral, double *cosPhi, RDGeom::Point3D r[4],
+    RDGeom::Point3D t[2], double d[2]) {
+  computeDihedral(static_cast<RDGeom::Point3D *>(pos[idx1]),
+    static_cast<RDGeom::Point3D *>(pos[idx2]),
+    static_cast<RDGeom::Point3D *>(pos[idx3]),
+    static_cast<RDGeom::Point3D *>(pos[idx4]), dihedral, cosPhi, r, t, d);
+}
+
+void computeDihedral(const double *pos, unsigned int idx1,
+    unsigned int idx2, unsigned int idx3, unsigned int idx4,
+    double *dihedral, double *cosPhi, RDGeom::Point3D r[4],
+    RDGeom::Point3D t[2], double d[2]) {
+  RDGeom::Point3D p1(pos[3 * idx1], pos[3 * idx1 + 1],
+                     pos[3 * idx1 + 2]);
+  RDGeom::Point3D p2(pos[3 * idx2], pos[3 * idx2 + 1],
+                     pos[3 * idx2 + 2]);
+  RDGeom::Point3D p3(pos[3 * idx3], pos[3 * idx3 + 1],
+                     pos[3 * idx3 + 2]);
+  RDGeom::Point3D p4(pos[3 * idx4], pos[3 * idx4 + 1],
+                     pos[3 * idx4 + 2]);
+  computeDihedral(&p1, &p2, &p3, &p4, dihedral, cosPhi, r, t, d);
+}
+
+void computeDihedral(const RDGeom::Point3D *p1, const RDGeom::Point3D *p2,
+    const RDGeom::Point3D *p3, const RDGeom::Point3D *p4,
+    double *dihedral, double *cosPhi, RDGeom::Point3D r[4],
+    RDGeom::Point3D t[2], double d[2]) {
+  RDGeom::Point3D rLocal[4];
```

Comment:
I think you need PRECONDITIONS on p1, p2, p3, and p4 making sure they aren't `nullptr`

Reply:
Good point, I will add them.